### PR TITLE
Feature/layer normalisation

### LINKED ISSUE #828: [INVESTIGATION]: Add layer normalisation to normalise hidden layers features
### What do you want to investigate?
The [onpolicy](https://github.com/marlbenchmark/on-policy/blob/79f6591882088a0f583f7a4bcba44041141f25f5/onpolicy/algorithms/utils/mlp.py#L19) repo implements feature normalisation by applying nn.Layernorm to the output of every linear layer. Currently we only normalise our input raw observations. We do not normlalise beyond the first layer.

### Definition of done
Add layer normalisation to every hidden layer of our networks in a similar fashion to the onpolicy repo.

### PR DESCRIPTION
**Pull request by ulricharmel**
## What?

- Added an option to apply layer normalisation to the hidden features of MLP networks.

## Why?

- To follow the implementation style of the [onpolicy repo](https://github.com/marlbenchmark/on-policy/blob/79f6591882088a0f583f7a4bcba44041141f25f5/onpolicy/algorithms/utils/mlp.py#L19)

## How?

- Clone haiku.nets.MLP and modified it to include a layer normalisation after every linear layer.
- Added a new network option to apply layer norm or not. Default is False.
- Modified every call to haiku.nets.MLP with the customise implementation. 

## Extra

- Closes #828

### CODE REVIEW

File: mava/utils/networks_utils.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
+"""Implementation an mlp module with layer normalisation"""
```
Comment (DriesSmit) [inline mava/utils/networks_utils.py:1]: Can you just please add the licence at the top of this file.

File: mava/utils/networks_utils.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+from typing import Any, Callable, Iterable, Optional
+
+import haiku as hk
+import jax
+import jax.numpy as jnp
+
+
+class MLP_NORM(hk.Module):
+    """A multi-layer perceptron module."""
```
Comment (DriesSmit) [inline mava/utils/networks_utils.py:12]: Can you just rename this.

File: mava/utils/networks_utils.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        output_sizes: Iterable[int],
+        w_init: Optional[hk.initializers.Initializer] = None,
+        b_init: Optional[hk.initializers.Initializer] = None,
+        with_bias: bool = True,
+        activation: Callable[[jnp.ndarray], jnp.ndarray] = jax.nn.relu,
+        activate_final: bool = False,
+        layer_norm: bool = False,
+        name: Optional[str] = None,
+    ):
+        """Constructs an MLP.
```
Comment (DriesSmit) [inline mava/utils/networks_utils.py:25]: Same here.

File: mava/utils/networks_utils.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        Args:
+            output_sizes: Sequence of layer sizes.
+            w_init: Initializer for :class:`~haiku.Linear` weights.
+            b_init: Initializer for :class:`~haiku.Linear` bias. Must be ``None`` if
+            ``with_bias=False``.
+            with_bias: Whether or not to apply a bias in each layer.
+            activation: Activation function to apply between :class:`~haiku.Linear`
+            layers. Defaults to ReLU.
+            activate_final: Whether or not to activate the final layer of the MLP.
+            layer_norm: apply layer normalisation hidden MLP layers.
```
Comment (DriesSmit) [inline mava/utils/networks_utils.py:35]: apply layer normalisation to the hidden MLP layers.

File: mava/utils/networks_utils.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+                # Only perform dropout if we are activating the output.
+                if dropout_rate is not None:
+                    out = hk.dropout(next(rng), dropout_rate, out)  # type: ignore
+                out = self.activation(out)
+            if self.layer_norm:
+                out = self.norms[i](out)
+
+        return out
+
+    def reverse(
```
Comment (DriesSmit) [inline mava/utils/networks_utils.py:113]: Is this function used internally when running our examples? Or is it required?

File: mava/utils/networks_utils.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        >>> rev = mlp.reverse()
+        >>> rev_mlp_out = rev(y)
+        >>> mlp_in.shape == rev_mlp_out.shape
+        True
+        Args:
+            activate_final: Whether the final layer of the MLP should be activated.
+            name: Optional name for the new module. The default name will be the name
+            of the current module prefixed with ``"reversed_"``.
+        Returns:
+            An MLP instance which is the reverse of the current instance. Note these
```
Comment (DriesSmit) [inline mava/utils/networks_utils.py:137]: Just rename MLP to MLP_NORM etc. This is still the old MLP text.

**Review by DriesSmit — COMMENTED** [review]
Thanks @ulricharmel :raised_hands: :slightly_smiling_face:  Hopefully this improves our system's performance :rocket: Just see my few comments.

File: mava/utils/networks_utils.py (hunk 5 above)
Comment (ulricharmel) [reply mava/utils/networks_utils.py]: No we don't use it. I have removed it. Also it was the one reducing the test coverage.

File: mava/utils/networks_utils.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+            layers.append(
+                hk.Linear(
+                    output_size=output_size,
+                    w_init=w_init,
+                    b_init=b_init,
+                    with_bias=with_bias,
+                    name="linear_%d" % index,
+                )
+            )
+            norms.append(
```
Comment (OmaymaMahjoub) [inline mava/utils/networks_utils.py:78]: I think it would be ideal if you first determine whether layer norm is True or False before calculating self.norms. Therefore in the case of non normalization you don't need to pass by this calculation.

Comment (ulricharmel) [reply mava/utils/networks_utils.py:78]: Thanks I will do that.

**Review by DriesSmit — DISMISSED** [review]
Thanks @ulricharmel :slightly_smiling_face:

**Review by OmaymaMahjoub — APPROVED** [review]
Thank you @ulricharmel :fire:

**Review by DriesSmit — APPROVED** [review]
Thanks @ulricharmel :raised_hands:
